### DISCUSSION TITLE
Feature Request: ergonomic way of doing a rolling mean with `min_periods`

### DISCUSSION THREAD

**Discussion by MarcoGorelli** [duckdb/duckdb#13784] [Ideas]
In feature engineering for time series forecasting, it's fairly common to create features using rolling means. For example, if I have a table with:
- one column with `store_id`
- one column with `date`
- one column with `sales`

then I could calculate a rolling mean of sales partitioned by `store_id` ordered by `date`

Assuming my data is already sorted by `'date'` within each `'store_id'`, in pandas I'd do
```python
df.groupby("store_id")['sales'].transform(lambda x: x.rolling(5).mean())
```

in Polars:
```python
df.with_columns(pl.col('sales').rolling_mean(5).over('store_id'))
```

In DuckDB, I found a [SO answer](https://stackoverflow.com/a/78953331/4451315) which suggests doing
```python
duckdb.sql("""
    from df
    select
        *,
        case
            when count(*) over rolling5 = 5 then 
                mean(sales) over rolling5
        end as sales_rolling
    window rolling5 as (
        partition by store_id 
        order by date
        rows between 4 preceding and current row
    )   
""")
```

This works but is super-complicated...in general I'm super-impressed with DuckDB, but rolling features are very difficult to write, and I think pandas/Polars are much easier to work with for this particular task

DuckDB has already introduced simplified sql syntax for pivot - might you consider also introducing some simplified syntax to make rolling aggregations easy (or at least easier) to write?

**Comment by soerenwolfers** [comment]
Personally, I don't find the duckdb solution complicated and it has the massive advantage over dataframe libraries that 90% of it isn't specific to duckdb (besides the other advantage that it doesn't require presorting). That's also why I'm against duckdb adding more ready-made solutions like `rolling_mean()` lightly -- there already is a very powerful, natural, and non-surprising way to achieve tweaks like `min_period` and it doesn't require new functions & syntax to be learned and maintained.

By the way, a minor tweak to your query:

```sql
case
when count(*) over rolling5 = 5 then 
      mean(sales) over rolling5
end as sales_rolling
```

could just be 
```sql
if(count() over rolling5 = 5, mean(sales) over rolling5, NULL) as sales_rolling
```

**Reply by MarcoGorelli** [reply]
thanks @soerenwolfers - those are valid points, I probably just need to practice my sql 😳

**Comment by hawkfish** [comment]
> Assuming my data is already sorted by 'date' within each 'store_id', in pandas I'd do

We are also discussing tracking partitioning and sorting metadata so we can avoid expensive steps like this.

Incidentally, another advantage of the DuckDB implementation is that we will compute the rolling means in parallel. Pandas is single-threaded. This may not matter much for small amounts of data but for larger volumes it will decrease the run time by a factor equal to the number of cores on your machine.
